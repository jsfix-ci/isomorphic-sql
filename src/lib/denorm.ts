import { and, csv } from './util/delims';
const deleted = 'deleted_at';

export const denorm = (table: string, columnNames: string[], primaryColumnNames: string[]) => {
    const selectDuplicitousRecords = [
        `SELECT COUNT(*), ${columnNames.join(csv)}`,
        `FROM ${table}`,
        `GROUP BY ${columnNames.join(csv)}`,
        `HAVING COUNT(*) > 1`,
    ].join(' ');

    const selectDuplicitousIds = [
        `SELECT ${columnNames.map((cn) => `dups.${cn}`).join(csv)}, ${table}.created_at, ${table}.id`,
        `FROM (${selectDuplicitousRecords}) AS dups`,
        `INNER JOIN ${table} ON ${[...primaryColumnNames.map((cn) => `${table}.${cn} = dups.${cn}`)].join(and)}`,
        `AND ((dups.${deleted} IS NULL AND ${table}.${deleted} IS NULL) OR ${table}.${deleted} = dups.${deleted})`,
        `ORDER BY ${table}.created_at ASC`,
    ].join(' ');

    return selectDuplicitousIds;
};