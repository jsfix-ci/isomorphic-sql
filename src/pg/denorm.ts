import { and, csv } from '../common/constant/delims';

export const denorm = (table: string, nullableColumnNames: string[], nonNullableColumnNames: string[], additionalColumnsToSelect: string[] = []) => {
    const columnNames = [...nullableColumnNames, nonNullableColumnNames];

    const selectDuplicitousRecords = [
        `SELECT COUNT(*), ${columnNames.join(csv)}`,
        `FROM ${table}`,
        `GROUP BY ${columnNames.join(csv)}`,
        `HAVING COUNT(*) > 1`,
    ].join(' ');

    const selectDuplicitousIds = [
        `SELECT ${columnNames.map((cn) => `dups.${cn}`).join(csv)}${additionalColumnsToSelect.length ? csv : ''} ${additionalColumnsToSelect.join(csv)}`,
        `FROM (${selectDuplicitousRecords}) AS dups`,
        `INNER JOIN ${table} ON ${[...nonNullableColumnNames.map((cn) => `${table}.${cn} = dups.${cn}`)].join(and)}`,
        and,
        nullableColumnNames.map(deleted => `((dups.${deleted} IS NULL AND ${table}.${deleted} IS NULL) OR ${table}.${deleted} = dups.${deleted})`).join(and),
        `ORDER BY ${table}.created_at ASC`,
    ].join(' ');

    return selectDuplicitousIds;
};