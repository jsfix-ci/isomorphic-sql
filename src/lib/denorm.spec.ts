import { denorm } from './denorm';

describe('denorm', () => {
    it('returns a sql string', () => {
        expect(denorm('person', ['person_type_id'], ['deleted_at'], ['first_name', 'last_name'])).toMatchSnapshot();
    });
});