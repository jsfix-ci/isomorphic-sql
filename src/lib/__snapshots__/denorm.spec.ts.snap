// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`denorm returns a sql string 1`] = `"SELECT dups.form_id, dups.quality_section_config_id, dups.deleted_at,  measure_id FROM (SELECT COUNT(*), form_id, quality_section_config_id, deleted_at FROM form_quality_section GROUP BY form_id, quality_section_config_id, deleted_at HAVING COUNT(*) > 1) AS dups INNER JOIN form_quality_section ON form_quality_section.deleted_at = dups.deleted_at  AND  ((dups.form_id IS NULL AND form_quality_section.form_id IS NULL) OR form_quality_section.form_id = dups.form_id) AND ((dups.quality_section_config_id IS NULL AND form_quality_section.quality_section_config_id IS NULL) OR form_quality_section.quality_section_config_id = dups.quality_section_config_id) ORDER BY form_quality_section.created_at ASC"`;
